pyface/data_view/data_models/row_table_data_model.py

```diff
@@ -0,0 +1,126 @@
+# is also available online at http://www.enthought.com/licenses/BSD.txt
+#
+# Thanks for using Enthought open source!
+""" A row-oriented data model implementation.
+
+This module provides a concrete implementation of a data model for the
+case of row-oriented data.
+"""
+from traits.api import Any, ComparisonMode, Instance, List, observe
+
+from pyface.data_view.abstract_data_model import (
+    AbstractDataModel, DataViewSetError
+)
+from pyface.data_view.index_manager import IntIndexManager
+from pyface.data_view.data_models.data_accessors import AbstractDataAccessor
+
+
+class RowTableDataModel(AbstractDataModel):
+    """ A data model that presents a list of objects as rows.
+    """
+
+    #: A sequence of objects to display in columns.
+    data = Any()
+
+    #: An object which describes how to map data for the row headers.
+    row_header_data = Instance(AbstractDataAccessor)
+
+    #: An object which describes how to map data for each column.
+    column_data = List(
+        Instance(AbstractDataAccessor),
+        comparison_mode=ComparisonMode.identity,
+    )
+
+    #: The index manager that helps convert toolkit indices to data view
+    #: indices.
+    index_manager = Instance(IntIndexManager, args=())
+
+    def get_column_count(self):
+        return len(self.column_data)
+
+    def can_have_children(self, row):
+        return len(row) == 0
+
+    def get_row_count(self, row):
+        if len(row) == 0:
+            return len(self.data)
+        else:
+            return 0
+
+    def get_value(self, row, column):
+        if len(column) == 0:
+            column_data = self.row_header_data
+        else:
+            column_data = self.column_data[column[0]]
+        if len(row) == 0:
+            return column_data.title
+        obj = self.data[row[0]]
+        return column_data.get_value(obj)
+
+    def can_set_value(self, row, column):
+        if len(row) == 0:
+            return False
+        if len(column) == 0:
+            column_data = self.row_header_data
+        else:
+            column_data = self.column_data[column[0]]
+        obj = self.data[row[0]]
+        return column_data.can_set_value(obj)
+
+    def set_value(self, row, column, value):
+        if len(row) == 0:
+            raise DataViewSetError("Can't set column titles.")
+        if len(column) == 0:
+            column_data = self.row_header_data
+        else:
+            column_data = self.column_data[column[0]]
+        obj = self.data[row[0]]
+        column_data.set_value(obj, value)
+        self.values_changed = (row, column, row, column)
+
+    def get_value_type(self, row, column):
+        if len(column) == 0:
+            column_data = self.row_header_data
+        else:
+            column_data = self.column_data[column[0]]
+        if len(row) == 0:
+            return column_data.title_type
+        return column_data.value_type
+
+    def _data_default(self):
+        return []
+
+    @observe('data')
```

Comment:
Without assuming a trait container type such as `List`, `data` can be any object conforming to the Sequence interface, such as a numpy array.  However the model won't have builtin support for value updates triggered by mutation on `data` if `data` is actually a TraitList.

Actually, I think it would be okay to replace this `@observe("data")` with `@observe(trait("data").list_items(optional=True))`. The `optional=True` will excuse the situation when `data` isn't a TraitList. But if `data` is a TraitList and is mutated, this model will fire `structure_changed` event. I think this is EEP2 and EEP3 proving their values here :)

Note: It would be naughty to use `"data.items"` because `"items"` is equivalent to `list_items(optional=True) | dict_items(optional=True) | set_items(optional=True) | trait("items", optional=True)`. We can't be sure if `data` is an instance of `HasTraits` that happens to have a trait named `"items"`; we don't want to be accidentally observing that trait for this update. And then dict and set aren't sequence so we are not expecting these being used as `data`.

Alternatively, we may need to instruct users to dynamically attach an observer if they need to observe mutation on `data` that is a trait list.

Reply:
I like `@observe(trait("data").list_items(optional=True))`, and it is indeed proving the value of the design when it allows you to do things you didn't know you needed to be able to do!

Reply:
But yes, it is part of the design that users may need to manually fire updates if they aren't working with Trait-y things.

Reply:
*sigh* Of course it doesn't work out of the box if you use a regular list, because it doesn't get copied into a `TraitsList`.

We need something better than the `List` trait in Traits.

Reply:
Ah you are referring to the `_data_default` returning a Python list. Maybe the default has to be an immutable sequence then? This would avoid causing confusion. If one needs a TraitList for `data`, assign one to it.

Reply:
No, I was referring to my attempts to write a test for the modified behaviour which weren't working because I was assigning a regular list expecting Traits to magically turn it into a trait list - just a brain-fart on my part.

However, returning an empty `TraitList` as the default may not be a bad idea.

Reply:
> However, returning an empty TraitList as the default may not be a bad idea.

Yes that would work too. What makes me very slightly worried is that we could be calling the constructor of `TraitList` directly, which to me is closer to playing with implementation details of Traits. The constructor of `TraitList` is not yet a published public API (I think). But maybe it was intended to be a published public API at the end, then this could be a way to test the API within ETS before making it more public?

Reply:
`TraitList` won't be part of the public API until `TraitHandler`s are fully removed, due to the name collision with the `TraitList` trait handler.  However it is something that I would like to eventually be public.

It may be better to spend the effort prototyping the "something better than the `List` trait," however.